lib/rom/sql/extensions/postgres/types/json.rb

```diff
@@ -16,16 +16,45 @@ module Types
           end
         end
 
-        JSON = Type('json') do
-          (SQL::Types::Array | SQL::Types::Hash)
-            .constructor(Sequel.method(:pg_json))
-            .meta(read: JSONRead)
-        end
+        if Sequel.respond_to?(:pg_json_wrap)
+          json_types = [
+            SQL::Types::Array,
+            SQL::Types::Hash,
+            SQL::Types::Integer,
+            SQL::Types::Float,
+            SQL::Types::String,
+            SQL::Types::Nil,
+            SQL::Types::True,
+            SQL::Types::False
+          ]
+
+          JSON = Type('json') do
+            json_cast_methods = Hash.new(:pg_json)
+            json_cast_methods.update(json_types.each_with_object({}) { |t, h| h[t.primitive] = :pg_json_wrap })
+
+            json_types
+              .reduce(:|)
+              .constructor { |value| Sequel.send(json_cast_methods[value.class], value) }
```

Comment:
IIRC `__send__` is more appropriate (if you wanted to use priv methods), otherwise it should be `public_send`.